src/Models/Workflow.php

```diff
@@ -189,6 +215,30 @@ public function asAdjacencyList(): array
         ])->all();
     }
 
+    private function isMarkedAsFinished(): bool
+    {
+        return ! is_null($this->finished_at);
+    }
+
+    private function markAsFinished(): void
+    {
+        $this->update(['finished_at' => Carbon::now()]);
+    }
+
+    private function markParentAsFinished(): void
+    {
+        optional($this->workflow, function (Workflow $workflow) {
+            if ($workflow->isMarkedAsFinished()) {
+                return;
+            }
+
+            if ($workflow->isFinished()) {
+                $workflow->markAsFinished();
+                $workflow->markParentAsFinished();
```

Comment:
This will recursively bubble up to workflow parents (and their parents), setting them to finished (if they are indeed finished).